metrax_example.ipynb

```diff
@@ -0,0 +1,526 @@
+        "    jitted_calculate = jax.jit(\n",
+        "        _calculate,\n",
+        "        in_shardings=(sharding_rule, sharding_rule, sharding_rule),\n",
+        "        out_shardings=None\n",
+        "    )\n",
+        "\n",
+        "    # The result is already a globally correct metric state, replicated on all devices.\n",
+        "    return jitted_calculate(sharded_predictions, sharded_labels, sharded_weights)\n",
+        "\n",
+        "\n",
+        "# Baseline: Single-Device (Direct)\n",
+        "@jax.jit\n",
+        "def calculate_aucpr_direct(predictions, labels, sample_weights):\n",
+        "  \"\"\"Computes AUCPR on the entire dataset on a single device.\"\"\"\n",
+        "  return metrax.AUCPR.from_model_output(\n",
+        "      predictions=predictions,\n",
+        "      labels=labels,\n",
+        "      sample_weights=sample_weights\n",
+        "  )\n",
+        "\n",
+        "\n",
+        "# --- 2. Execution and Verification ---\n",
+        "print(\"\\nRunning all three AUCPR calculation methods...\")\n",
+        "\n",
+        "# Execute each of the three methods.\n",
+        "state_pmap = calculate_aucpr_pmap(predictions, labels, sample_weights)\n",
+        "state_mesh = calculate_aucpr_mesh(predictions, labels, sample_weights)\n",
+        "state_direct = calculate_aucpr_direct(predictions, labels, sample_weights)\n",
+        "\n",
+        "# Compute the final values from the metric states.\n",
+        "result_pmap = state_pmap.compute()\n",
+        "result_mesh = state_mesh.compute()\n",
+        "result_direct = state_direct.compute()\n",
+        "\n",
+        "# Ensure all computations are finished before verifying.\n",
+        "result_pmap.block_until_ready()\n",
+        "result_mesh.block_until_ready()\n",
+        "result_direct.block_until_ready()\n",
+        "\n",
+        "# Verify that all results are numerically identical.\n",
+        "assert np.allclose(result_pmap, result_direct, rtol=1e-6)\n",
+        "assert np.allclose(result_mesh, result_direct, rtol=1e-6)\n",
+        "\n",
+        "\n",
+        "# --- 3. Display Results ---\n",
+        "print(\"\\n\" + \"=\"*60)\n",
+        "print(\"          Comparison of Multi-Device AUCPR Calculations\")\n",
+        "print(\"=\"*60)\n",
+        "print(f\"{'Method':<35} {'AUCPR Value'}\")\n",
+        "print(\"-\" * 60)\n",
+        "print(f\"{'Method 1: pmap':<35} {result_pmap}\")\n",
+        "print(f\"{'Method 2: jit + Mesh':<35} {result_mesh}\")\n",
+        "print(f\"{'Baseline: Direct Single-Device':<35} {result_direct}\")\n",
+        "print(\"=\"*60)\n",
+        "print(\"\\n✅ Verification successful: All three methods yield identical results.\")"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "C3YWS1_x19DJ"
+      },
+      "source": [
+        "## 🧠 Advanced Use: Multi-Host Environments\n",
```

Comment:
I like the shoutout here, for anything further I think it makes sense for this to be in its own Colab